icu4c/source/common/locbased.h

```diff
@@ -41,13 +42,13 @@ class U_COMMON_API LocaleBased : public UMemory {
      * Construct a LocaleBased wrapper around the two pointers.  These
      * will be aliased for the lifetime of this object.
      */
-    inline LocaleBased(char* validAlias, char* actualAlias);
+    inline LocaleBased(CharString* validAlias, CharString* actualAlias);
 
     /**
      * Construct a LocaleBased wrapper around the two const pointers.
      * These will be aliased for the lifetime of this object.
      */
-    inline LocaleBased(const char* validAlias, const char* actualAlias);
+    inline LocaleBased(const CharString* validAlias, const CharString* actualAlias);
```

Comment:
This, and the implementation which casts away const, looks wrong.
None of the classes that use LocaleBased should have `const CharString`.

Please try to remove this constructor completely.